theano/compile/builders.py

```diff
@@ -121,14 +144,110 @@ def make_node(self, *inputs):
+
+        _c_code = "{\n"
+        self.nodenames = ["%(nodename)s_" + ('subnode%i' % j)
+                          for j, n in enumerate(self.fgraph.toposort())]
+
+        i = 0
+        for j, node in enumerate(self.fgraph.toposort()):
+            for output in node.outputs:
+                if output not in subd:
+                    i += 1
+                    name = "V%%(id)s_tmp%i" % i
+                    subd[output] = name
+                    _c_code += "%s %s;\n" % (
+                            output.type.dtype_specs()[1], name)
+            s = node.op.c_code(
+                node,
+                self.nodenames[j],
+                [subd[input] for input in node.inputs],
+                [subd[output] for output in node.outputs],
+                dict(fail="%(fail)s", id="%%(id)s_%i" % j))
+            _c_code += s
+            _c_code += "\n"
+        _c_code += "}\n"
+
+        for char in 'idflxsc':
+            _c_code = re.sub('\%'+char, "%%"+char, _c_code)
+
+        self._c_code = _c_code
+
+    def c_code(self, node, nodename, inames, onames, sub):
+        dic = dict(chain(izip(("i%i" % i for i in xrange(len(inames))), inames),
+                       izip(("o%i" % i for i in xrange(len(onames))),
+                            onames)), **sub)
+        dic['nodename'] = nodename
+        if 'id' not in sub:
+            # The use of a dummy id is safe as the code is in a separate block.
+            # It won't generate conflicting variable name.
+            dic['id'] = '_DUMMY_ID_'
+
+        return self._c_code% dic
+
+    def c_code_cache_version(self):
+        rval = [3]
+        for x in self.fgraph.toposort():
+            xv = x.op.c_code_cache_version()
+            if xv:
+                rval.append(xv)
+            else:
+                return ()
+        return tuple(rval)
+
+    def c_support_code(self):
+        rval = []
+        for subnode in self.fgraph.toposort():
+            try:
+                rval.append(subnode.op.c_support_code().strip())
+            except gof.utils.MethodNotDefined:
+                pass
+        # remove duplicate code blocks
+        return "\n".join(sorted(set(rval)))
+
+    def c_support_code_apply(self, node, name):
+        rval = []
+        for subnode, subnodename in zip(self.fgraph.toposort(), self.nodenames):
+            try:
+                subnode_support_code = subnode.op.c_support_code_apply(
+                    subnode,
+                    subnodename % dict(nodename=name))
+                if subnode_support_code:
+                    rval.append(subnode_support_code)
+            except gof.utils.MethodNotDefined:
+                pass
+        # there should be no need to remove duplicate code blocks because
+        # each block should have been specialized for the given nodename.
+        # Any block that isn't specialized should be returned via
+        # c_support_code instead of c_support_code_apply.
+        return "\n".join(rval)
+
```

Comment:
There are lots of methods that help C code generation other than c_support_code() and c_support_code_apply().

See http://deeplearning.net/software/theano/extending/cop.html#what-needs-to-be-defined to get the list.